src/System.Net.Primitives/src/System/Net/IPAddressPal.Unix.cs

```diff
@@ -0,0 +1,201 @@
+
+        public static unsafe uint Ipv4AddressToString(byte[] address, StringBuilder buffer)
+        {
+            Debug.Assert(address != null);
+            Debug.Assert(address.Length == IPAddressParser.IPv4AddressBytes);
+            Debug.Assert(buffer != null);
+            Debug.Assert(buffer.Capacity >= IPAddressParser.INET_ADDRSTRLEN);
+
+            var sockaddr = new Interop.libc.sockaddr_in {
+                sin_family = Interop.libc.AF_INET,
+                sin_port = 0
+            };
+
+#if BIGENDIAN
+            sockaddr.sin_addr.s_addr = unchecked((uint)(address[3] | (address[2] << 8) | (address[1] << 16) | (address[0] << 24)));
+#else
+            sockaddr.sin_addr.s_addr = unchecked((uint)(address[0] | (address[1] << 8) | (address[2] << 16) | (address[3] << 24)));
+#endif
+            int err = Interop.libc.getnameinfo((Interop.libc.sockaddr*)&sockaddr, (uint)sizeof(Interop.libc.sockaddr_in), buffer, (uint)buffer.Capacity, null, 0, Interop.libc.NI_NUMERICHOST);
+            return unchecked((uint)err);
+        }
+
+        public static unsafe uint Ipv6AddressToString(byte[] address, uint scopeId, StringBuilder buffer)
+        {
+            Debug.Assert(address != null);
+            Debug.Assert(address.Length == IPAddressParser.IPv6AddressBytes);
+            Debug.Assert(buffer != null);
+            Debug.Assert(buffer.Capacity >= IPAddressParser.INET6_ADDRSTRLEN);
+
+            var sockaddr = new Interop.libc.sockaddr_in6 {
+                sin6_family = Interop.libc.AF_INET6,
+                sin6_port = 0,
+                sin6_scope_id = scopeId
+            };
+
+            Debug.Assert(sizeof(Interop.libc.in6_addr) == IPAddressParser.IPv6AddressBytes);
+            for (int i = 0; i < IPAddressParser.IPv6AddressBytes; i++)
+            {
+                sockaddr.sin6_addr.s6_addr[i] = address[i];
+            }
+
+            int err = Interop.libc.getnameinfo((Interop.libc.sockaddr*)&sockaddr, (uint)sizeof(Interop.libc.sockaddr_in6), buffer, (uint)buffer.Capacity, null, 0, Interop.libc.NI_NUMERICHOST);
+            return unchecked((uint)err);
+        }
+
+        public static unsafe uint Ipv4StringToAddress(string ipString, byte[] bytes, out ushort port)
+        {
+            Debug.Assert(ipString != null);
+            Debug.Assert(bytes != null);
+            Debug.Assert(bytes.Length == IPAddressParser.IPv4AddressBytes);
+
+			port = 0;
+
+            var hints = new Interop.libc.addrinfo {
+                ai_flags = Interop.libc.AI_NUMERICHOST | Interop.libc.AI_NUMERICSERV,
+                ai_family = Interop.libc.AF_INET,
+                ai_socktype = 0,
+                ai_protocol = 0
+            };
+
+            Interop.libc.addrinfo* addrinfo = null;
+            int err = Interop.libc.getaddrinfo(ipString, null, &hints, &addrinfo);
+            if (err != 0)
+            {
+                Debug.Assert(addrinfo == null);
+                return unchecked((uint)err);
```

Comment:
@pgavlin Error handling is a complicated issue. Exceptions are part of the networking patterns (e.g. retry on connection failure) and so we need to be extra careful to create the same behaviors in all x-plat implementations.

What I propose (similar to what you said above) is that exceptions always match between architectures: e.g. for Sockets, I planned to use the platform agnostic SocketError code and have an adapter for each platform. Unit tests should exist to pin down the error behavior in all platforms. (An example exists for WebSockets.Client where the same tests have been added to Desktop and corefx for at least some of the error patterns.) 

At the same time, as an app-developer, I would still expect that I can eventually get the actual OS code as it may relay more information particular to the platform that isn't easily correlated to Windows. For that, I propose we use Exception.HResult or attach a platform specific inner exception so that the code to user-readable text is also exposed.

@terrajobst @weshaggard @ericstj @davidsh : What is your opinion on this?